runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/JobGroupTest.java

```diff
@@ -1293,6 +1309,88 @@ public IStatus run(IProgressMonitor monitor) {
 		assertTrue("2.0", numShouldCancelCalled[0] < NUM_JOBS_LIMIT + 10);
 	}
 
+	/**
+	 * Tested scenario: - JobGroup.shouldCancel returns true after a single job is
+	 * canceled.
+	 *
+	 * Expected result: The remaining jobs are canceled in a reasonable time after
+	 * the shouldCancel method of the JobGroup returns true. Some jobs might even
+	 * have to be explicitly asked to cancel their execution via the method
+	 * "canceling"
+	 */
+	public void testShouldCancel_5() {
+		final int MAX_THREADS = 10;
+
+		// make sure some jobs haven't yet been queued by the time the cancelation takes
+		// place
+		final int NUM_JOBS_LIMIT = MAX_THREADS + 100;
+		final int NUM_JOBS = NUM_JOBS_LIMIT + 1000;
+
+		// counters
+		final int countJobsRan[] = { 0 };
+		final int countJobsCanceled[] = { 0 };
+		final TestBarrier2 barrier = new TestBarrier2();
+
+		final JobGroup jobGroup = new JobGroup("JobGroup", MAX_THREADS, NUM_JOBS) {
+			@Override
+			protected boolean shouldCancel(IStatus lastCompletedJobResult, int numberOfFailedJobs,
+					int numberOfCanceledJobs) {
+				// cancel the group as soon as one job in it is canceled
+				return numberOfCanceledJobs > 0;
+			}
+		};
+		for (int i = 0; i < NUM_JOBS; i++) {
+			final int jobNumber = i;
+
+			Job job = new TestJob("TestJob", 1, 1) {
+
+				private boolean canceledExplicitly;
+				@Override
+				public IStatus run(IProgressMonitor monitor) {
+					barrier.waitForStatus(TestBarrier2.STATUS_START);
+
+					if (canceledExplicitly || jobNumber == NUM_JOBS_LIMIT) {
+						// after this cancelation, no more jobs will be scheduled and the running jobs
+						// should be explicitly asked to cancel their execution
+						return Status.CANCEL_STATUS;
+					}
+
+					super.run(monitor);
+					countJobsRan[0]++; // this job ran
+
+					return new Status(IStatus.INFO, "org.eclipse.core.jobs", "Job " + jobNumber);
+				}
+
+				@Override
+				protected void canceling() {
+					// this job was explicitly asked to cancel its execution because it was already
+					// running (or about to)
+					canceledExplicitly = true;
+					countJobsCanceled[0]++;
+				}
+			};
+
+			job.setJobGroup(jobGroup);
+			job.schedule();
+		}
+
+		// Allow the jobs to proceed to run.
+		barrier.setStatus(TestBarrier2.STATUS_START);
+		waitForCompletion(jobGroup);
+
+		// Some jobs might have already been running by the time the cancelation took
+		// place
+		assertTrue("Too many jobs ran: " + countJobsRan[0], //
+				countJobsRan[0] <= NUM_JOBS_LIMIT + MAX_THREADS);
```

Comment:
Please use hamcrest with a proper lessThanOrEqualTo comparison.

Reply:
There is not memory barrier on `countJobsRan`, so this comparison can succeed although it shouldn't.